### DISCUSSION TITLE
Weird video flipping?

### DISCUSSION THREAD

**Discussion by eringiglio** [talmolab/sleap#1056] [Help!] [answered]
For whatever reason, some of my videos on my desktop machine display in SLEAP are showing up as though they've been vertically and horizontally inverted, while displaying properly on my laptop. The video files themselves are not inverted (I've checked through opening the same files in a video editor) and the files are hosted on google drive through the cloud. 

Both machines are on Windows 11 OS. Do you have any idea what might be causing this inversion or ideas as to how I can make it stop?

**Comment by talmo** [comment]
Hi @eringiglio,
I don't think we've seen this before! Do you sharing an example video file with us where this happens? You can email a link to `talmo@salk.edu`.

Also have you tried [reencoding](https://sleap.ai/help.html#does-my-data-need-to-be-in-a-particular-format) the video to see if it still happens?

Thanks!

**Comment by eringiglio** [comment]
I have tried re-encoding the video! It doesn't seem to help. It's the most bizarre thing. Do you think it could be something in the engine SLEAP uses to load video? I've shared an example video with you, but let me know if you'd like me to share the example SLP project file too.

**Comment by roomrys** [comment]
Piping in here:

There is [this post](https://support.google.com/drive/thread/3895385/video-from-google-drive-are-upside-down-once-downloaded?hl=en) about Google Drive videos being flipped once downloaded which stems from the [media player not being able to handle the rotation metadata](https://blog.addpipe.com/mp4-rotation-metadata-in-mobile-video-files/).

**Comment by eringiglio** [comment]
Oh, thank you! That looks like a potentially involved aspect--but why would
the SLEAP media player be different (or at least differently able to handle
the rotation metadata) between two computers? Is there a way to look at how
SLEAP is handling that metadata within a given installation?

On Mon, Nov 21, 2022 at 2:47 PM Liezl Maree ***@***.***>
wrote:

> Piping in here:
>
> There is this post
> <https://support.google.com/drive/thread/3895385/video-from-google-drive-are-upside-down-once-downloaded?hl=en>
> about Google Drive videos being flipped once downloaded which stems from
> the media player not being able to handle the rotation metadata
> <https://blog.addpipe.com/mp4-rotation-metadata-in-mobile-video-files/>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/talmolab/sleap/issues/1040#issuecomment-1322621825>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB55UJSSV2BP7RNCDAMBZHDWJPNWNANCNFSM6AAAAAAR44JU6U>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Erin Giglio, PhD.
Postdoctoral Associate
University of Minnesota
(770)-653-0022

**Comment by talmo** [comment, marked as answer]
Okidokie, I played around with the video you sent.

The issue is that the original video has some rotation metadata that is partially supported depending on where we're reading it, as @roomrys suggested. SLEAP uses opencv and ffmpeg under the hood for reading videos, and this usually will support rotation, but not on every environment/platform.

We can definitely get it encoded a bit more efficiently and fix the rotation by using this command:

```
ffmpeg -y -i bacon1_drug2.22.MP4 -an -c:v libx264 -preset superfast -pix_fmt yuv420p -crf 29 -metadata:s:v:0 rotate=0 bacon1_drug2.22.sf.mp4
```

The important parts:
- `-an`: This removes the extraneous audio track that's stored in the file. I assume you don't need this, but if you're using it somehow, you can just take out that bit of the command.
- `-crf 29`: Since you have long videos, you'll probably want to push on the compression a little bit. The CRF (compression rate factor) increases the amount of compression as the value increases. I used 29 and got the file size down from ~19 GB to ~1 GB with little loss in quality. You could experiment with 30-32 if you want to push it even further for space efficiency.
- `-metadata:s:v:0 rotate=0`: This sets the rotation metadata to 0 degrees, meaning it should not rotate the pixels even if the software that's reading it supports that.

I'm not sure the best way to test this other than to just try it across the different systems you were on, but I'll send you the reencoded file back over email and you can let us know if that works for you.

**Comment by eringiglio** [comment]
I've just checked, and that's done it! Thank you so much; I'll re-encode the video files and that should work well moving forward. I really appreciate the detailed workover of the ffmpeg commands, too!
